website/project/model.py

```diff
@@ -2829,188 +2911,313 @@ def to_json(self):
-            )
-            # Remove any embargoes associated with the registration
-            if parent_registration.embargo_end_date or parent_registration.pending_embargo:
-                parent_registration.embargo.state = self.CANCELLED
-                parent_registration.registered_from.add_log(
-                    action=NodeLog.EMBARGO_CANCELLED,
-                    params={
-                        'node': parent_registration._id,
-                        'embargo_id': parent_registration.embargo._id,
-                    },
-                    auth=Auth(user),
-                )
-                parent_registration.embargo.save()
-            # Ensure retracted registration is public
-            if not parent_registration.is_public:
-                parent_registration.set_privacy('public')
-            parent_registration.update_search()
-            # Retraction status is inherited from the root project, so we
-            # need to recursively update search for every descendant node
-            # so that retracted subrojects/components don't appear in search
-            for node in parent_registration.get_descendants_recursive():
-                node.update_search()
-
-
-def validate_embargo_state(value):
-    acceptable_states = [
-        Embargo.UNAPPROVED, Embargo.ACTIVE, Embargo.CANCELLED, Embargo.COMPLETED
-    ]
-    if value not in acceptable_states:
-        raise ValidationValueError('Invalid embargo state assignment.')
-    return True
+    def reject(self, user, token):
+        """Cancels sanction if user is admin and token verifies."""
+        try:
+            if self.approval_state[user._id]['rejection_token'] != token:
+                raise InvalidSanctionRejectionToken(self.REJECTION_INVALID_TOKEN_MESSAGE.format(DISPLAY_NAME=self.DISPLAY_NAME))
+        except KeyError:
+            raise PermissionsError(self.REJECTION_NOT_AUTHORIZED_MESSAEGE.format(DISPLAY_NAME=self.DISPLAY_NAME))
+        self.state = Sanction.REJECTED
+        self._on_reject(user, token)
 
+    def _notify_authorizer(self, user):
+        pass
+
+    def _notify_non_authorizer(self, user):
+        pass
+
+    def ask(self, group):
+        for contrib in group:
+            if contrib._id in self.approval_state:
+                self._notify_authorizer(contrib)
+            else:
+                self._notify_non_authorizer(contrib)
+
+class EmailApprovableSanction(Sanction):
+
+    AUTHORIZER_NOTIFY_EMAIL_TEMPLATE = None
+    NON_AUTHORIZER_NOTIFY_EMAIL_TEMPLATE = None
+
+    VIEW_URL_TEMPLATE = ''
+    APPROVE_URL_TEMPLATE = ''
+    REJECT_URL_TEMPLATE = ''
+
+    # Store a persistant copy of urls for use when needed outside of a request context.
+    # This field gets automagically updated whenever models approval_state is modified
+    # and the model is saved
+    # {
+    #   'abcde': {
+    #     'approve': [APPROVAL_URL],
+    #     'reject': [REJECT_URL],
+    #   }
+    # }
+    stashed_urls = fields.DictionaryField()
```

Comment:
Is the default for DictionaryFields an empty dict or `None`? You might want to explicitly set it to `dict`  to avoid a possible  TypeError in `add_authorizer`.